### ISSUE TITLE
Support for OV5640

### ISSUE THREAD

**Issue by CountParadox** [jomjol/AI-on-the-edge-device#2162]
### The Feature

Hi guys, 

I have this module:
https://www.aliexpress.com/item/1005004521039608.html?aem_p4p_detail=20230311200737107046978822230007724708&algo_exp_id=973ada50-63fc-40e0-91c3-87db4e401cd5-2&pdp_ext_f=%7B%22sku_id%22%3A%2212000031106421469%22%7D&pdp_npi=3%40dis%21AUD%2136.52%2123.0%21%21%21%21%21%40211bf3f716785940569804937d06c0%2112000031106421469%21sea%21AU%212214764457&curPageLogUid=fhcmfz3jOSwK&ad_pvid=20230311200737107046978822230007724708_3&ad_pvid=20230311200737107046978822230007724708_3

I think the inclusion of USB-C is very convenient. 
I think it should be quite simple to support, maybe just changing the pinouts defined in the code? 
I was able to get it working with ESPhome very easily by doing this.

What are others opinion? 

Thanks

**Comment by caco3** [comment]
Did you try if it works?

AFAIK there is no pinout configuration, we simply use the ESP library which provides the support for the different cameras.
See https://github.com/espressif/esp32-camera/tree/5c8349f4cf169c8a61283e0da9b8cff10994d3f3

**Comment by CountParadox** [comment]
@caco3  looks like its supported by the underlying library, but I need to be able to define the pins its connected to, I forked and had a go at this but I am pretty bad at code...
The increased resolution may not be an issue given that its not running video, just stills ?

**Comment by Slider0007** [comment]
It could possibly work with some modifications:

Camera pinouts are defined here: https://github.com/jomjol/AI-on-the-edge-device/blob/933215c1164d8b2b816a87ad42df185563939b1b/code/include/defines.h#L201

Additionally this flag needs to set: https://github.com/jomjol/AI-on-the-edge-device/blob/933215c1164d8b2b816a87ad42df185563939b1b/code/sdkconfig.defaults#L135

Be aware, that for other peripherals pins needs to be adpated as well:
- SD card
https://github.com/jomjol/AI-on-the-edge-device/blob/933215c1164d8b2b816a87ad42df185563939b1b/code/main/main.cpp#L117
- LED
- Flashlight (seems not available on this board)
https://github.com/jomjol/AI-on-the-edge-device/blob/933215c1164d8b2b816a87ad42df185563939b1b/code/include/defines.h#L62

- The resolution needs to be set max. to 640x480 to ensure that available memory is sufficient.

Only testing can tell you, if more modifications are necessary.
This is just what came in mind to me...

**Comment by commentsonly** [comment]
I'd say according to the table shown earlier there's no flashlight soldered but the table states that pin GPIO25 is the right one, as long as not attaching a TFT which would occupy that IO port.

**Comment by caco3** [comment]
keep in mind that we don't have RAM to support a higher camera resolution!

**Comment by AhmedThabit** [comment]
@CountParadox could you figure out a solution for this camera?

**Comment by CountParadox** [comment]
Negative, I just purchased a supported board

**Comment by AhmedThabit** [comment]
can you recommend a one ?

**Comment by jasaw** [comment]
I managed to get my OV5640 to work on my local build. I just need to enable ov5640 in SDK config, added guards around ov2640 specific code because they are not compatible to ov5640. Zoom mode is also working on ov5640 after very slight modification in the zoom handling code in `ClassControllCamera.cpp`.

@caco3 I would create a PR if I can get the html working but create reference image html on rolling branch appears to be broken? What is the state of rolling branch now? Sorry I haven't looked into this for a few months.

**Comment by caco3** [comment]
> @caco3 I would create a PR if I can get the html working but create reference image html on rolling branch appears to be broken? What is the state of rolling branch now? Sorry I haven't looked into this for a few months.

I am not aware of any recent changes on the Web UI. Can you try the latest rolling and describe your problem a bit more in detail?

**Comment by jasaw** [comment]
>I am not aware of any recent changes on the Web UI. Can you try the latest rolling and describe your problem a bit more in detail?

@caco3 sorry, turns out that it was just my web browser caching the previous version of the web interface causing problem with the new firmware on rolling branch. I just need to clear my browser cache.

I've created a PR to support OV5640.
https://github.com/jomjol/AI-on-the-edge-device/pull/3063

**Comment by jasaw** [comment]
Has anyone tried the [OV5640 camera with autofocus](https://www.aliexpress.com/item/1005004072836395.html)?

Just need to solder 2 wires on the ESP32cam CSI connector to power the autofocus control. The last 2 pins are not connected at the moment.
![ESP32-CAM-AI-Thinker-schematic-diagram](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/975a9235-f448-4c12-a8c8-bccf3449aaa5)

This is the OV5640 with autofocus pinout.
![ov5640af_pinout](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/64a7e079-b155-4d3a-b3fe-753e8dae4fab)

Here's an open source project that demonstrates the autofocus feature.
https://github.com/0015/ESP32-OV5640-AF

**Comment by jasaw** [comment]
The [esp32cam OV5640 autofocus aliexpress listing](https://www.aliexpress.com/item/1005004521039608.html) that I mentioned above does indeed have the autofocus AF-VCC and AF-GND connected on the CSI connector, so no soldering required.

![esp32cam_plus_ov5640_af](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/6c3250b8-eb4e-402f-a51d-eb93f0fece96)

**Comment by SybexX** [comment]
https://github.com/0015/ESP32-OV5640-AF/blob/main/README.md

**Comment by jasaw** [comment]
@SybexX Just a quick update on my progress on autofocus:
- I have seen 3 variants of OV5640 with AF: waveshare (AF-VDD on pin24), Adafruit (AF-VDD on pin23), AliExpress (AF-VDD pin23 & AF-GND pin24). I'm using the AliExpress variant.
- Autofocus works beautifully when I power AF-VDD with 2.8V and AF-GND with GND. I've shared my hardware mod photo below.
- I am using the OV5640 autofocus microcode from https://github.com/0015/ESP32-OV5640-AF but had to modify all the other functions for AF to work.
- [ov5640-firmware-user-guide.pdf](https://github.com/jomjol/AI-on-the-edge-device/files/15289609/ov5640-firmware-user-guide.pdf) describes step by step exactly how to get AF to work. Basically the steps are, load AF microcode to initialize, then on each flow cycle, set single shot AF mode, wait for it to get focus by polling focus status, capture image for the flow cycle, release AF.
- I have yet to figure out how to manually set the focus level. I think we have to manipulate registers 0x3602 and 0x3603.


![20240522_125703](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/39575165-251a-42a9-b7d0-09bad8fee77d)

**Comment by alekseyprisekin** [comment]
@jasaw. Like on this photo, does it working? Did u checked this in real? Or only in the serial monitor?
And where u connect cable, i cant understand?
![Снимок экрана 2024-09-17 104006](https://github.com/user-attachments/assets/d0a67511-04fb-4f73-a9f7-470227923677)

**Comment by jasaw** [comment]
@alekseyprisekin Yes auto focus works (I checked the captured image), but requires additional code to enable auto focus. I haven't shared my auto focus code on github yet because it's still work in progress.

On my photo, the white wire is soldered on U1 pin2, which is the 2.8V LDO output.

There are 3 different variants of OV5640 with auto focus and they require 2.8V and/or GND to be connected to different pins.

**Comment by alekseyprisekin** [comment]
@jasaw 
This basic code for camerawebserver (edit for autofocus), don't work with this version of ov5640?

**Comment by jasaw** [comment]
@alekseyprisekin I have not tried the project that you mentioned, but I started my auto focus code based on that project and had to make changes for auto focus to work on my camera. There are 3 different datasheets floating around the Internet, but I got time to investigate further.

**Comment by jasaw** [comment]
@alekseyprisekin Focus support is implemented on this PR if you want to try it out.
https://github.com/jomjol/AI-on-the-edge-device/pull/3504
